### PR TITLE
Event loop

### PR DESCRIPTION
**Pull request by ksqsf**
To test, observe whether `schedule`-d actions actually happen by adding a debug print or a breakpoint. Some instances:

- `RimeEngine::rimeNotificationHandler`
- `PinyinEngine::populateConfig`

Notifications can't be tested (just yet) because that PR depends on some changes made by this PR.

## For prosperity...

There are two important threads: the UI thread (GCD main thread) and the fcitx thread. Any code after this PR should adhere to the following rules:

+ UI operations can only be issued in the UI thread. Use `DispatchQueue.main.async`.
+ Fcitx and its addons should only be modified in the fcitx thread. Use the `with_fcitx` helper function.

### CODE REVIEW

File: src/fcitx.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    void exec();
+    void exit();
+    void schedule(std::function<void()>);
+
+    fcitx::AddonManager &addonMgr();
+    fcitx::AddonInstance *addon(const std::string &name);
+    fcitx::MacosFrontend *macosfrontend();
+
+private:
+    friend void start_fcitx_thread();
```
Comment (eagleoflqj) [inline src/fcitx.h:29]: Not needed.

File: src/fcitx.cpp [hunk 2]
```diff
@@ -78,24 +89,90 @@ void start_fcitx() {
...
+}
+
+fcitx::AddonManager &Fcitx::addonMgr() { return instance_->addonManager(); }
+
+fcitx::AddonInstance *Fcitx::addon(const std::string &name) {
+    return addonMgr().addon(name);
+}
+
+fcitx::MacosFrontend *Fcitx::macosfrontend() {
+    return dynamic_cast<fcitx::MacosFrontend *>(addon("macosfrontend"));
```
Comment (eagleoflqj) [inline src/fcitx.cpp:110]: `macosfrontend` is frequently used, but every call to `addon` queries a map.
https://github.com/fcitx/fcitx5/blob/b56322a85dbc06d74dc855770971d642f69b662e/src/lib/fcitx/addonmanager.cpp#L67
The function could be kept, but the result of `addon("macosfrontend")` should be cached.

**Review by eagleoflqj — CHANGES_REQUESTED** [review]

**Review by eagleoflqj — APPROVED** [review]
Tested with lldb. `RimeEngine::rimeNotificationHandler` is hit on Shift.
Thanks for separating threads.

**Comment by ksqsf** [comment]
Note: `showPreedit` and `commit` were wrapped in `DispatchQueue.main.async` at first (https://github.com/fcitx-contrib/fcitx5-macos/pull/16/commits/e1f3049cf8cc88a8ecb774a1d5424d1ff138268a), but due to issues with iTerm2, it was reverted (https://github.com/fcitx-contrib/fcitx5-macos/pull/16/commits/a37816150329876a62c804c60915cfa5908167f8). This is harmless according to my local tests, but further investigation can be helpful.

**Pull request merged**